[MultiThreading] Fix Livers scenes crash

fspadoni opened the pull request:
CHANGE: AdvanceTimer::Id::IdFactory is no more a singleton static class but  a thread local class to prevent different threads to share the same instance of the class.

FIX: DefaultTaskScheduler registration to TaskScheduler factory is forced to be executed after the TaskScheduler static initialization.




______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

fspadoni committed:
[MultiThreading] Change  AdvanceTimer::Id::IdFactory getInstance() static function to instantiate and return thread local variables. FIX DefaultTaskScheduler registration to TaskScheduler factory

guparan commented:
defrost-macos1011-0 (Defrost machine): AppleClang 8.0.0.8000042
sofa-ci-dev-macos109-0 (VM): AppleClang 6.0.0.6000051

The machine on which a build was done is printed at the beginning of the log.
Example: "Building remotely on sofa-ci-dev-macos109-0" in https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-792/CI_CONFIG=mac_clang-3.5,CI_PLUGINS=options,CI_TYPE=release/1/console

fspadoni committed:
[MultiThreading] Fix local static var initialization

fspadoni committed:
[MultiThreading] replace thread_specific_ptr with c++11 thread_local

fspadoni commented:
[ci-build][with-scene-tests]

fspadoni committed:
[MultiThreading] Recover the thread_specific_ptr version of getInstance()

fspadoni committed:
[MultiThreading] replace thread_specific_ptr with SOFA_THREAD_SPECIFIC_PTR macro

fspadoni commented:
[ci-build][with-scene-tests]

guparan commented:
Hey @fspadoni, no need to manually trigger the builds, they are automatically triggered on push :wink:

fspadoni commented:
@guparan OK, and is there a way to interrupt a build and start a new one right away?  .. in case I realized I committed something wrong

fspadoni committed:
[MultiThreading] remove template notation

guparan commented:
@fspadoni Just ask on Gitter and someone with enough privilege (me or Damien I guess) will do it for you :-)
The thing is that if you manually start a build right after a push, 2 builds will be racing for the Dashboard (same line for both) so the results may be corrupted.

fspadoni commented:
@guparan Thanks for the explanation. 
Some of the yesterday build failed were probably caused for the concurrent build. 
May I run one clean build today to be sure the build failed on mac_clang was not affected by that problem ?

guparan commented:
Sure, I do it right away: [ci-build][with-scene-tests]

guparan commented on applications/plugins/MultiThreading/src/DefaultTaskScheduler.cpp:
```diff
@@ -21,7 +21,7 @@ namespace sofa
         std::map< std::thread::id, WorkerThread*> TaskSchedulerDefault::_threads;
         
-        const bool TaskSchedulerDefault::isRegistered = TaskScheduler::registerScheduler(TaskSchedulerDefault::name(), &TaskSchedulerDefault::create);
+//        const bool TaskSchedulerDefault::isRegistered = TaskScheduler::registerScheduler(TaskSchedulerDefault::name(), &TaskSchedulerDefault::create);
```
To be removed if not used.

fspadoni committed:
[MultiThreading] Changes required for the review of PR 792.
Rename TaskSchedulerDefault to DefaultTaskScheduler. Remove commented code. Restore the const specifier for DefaultTaskScheduler::isRegistered static vars

guparan commented:
@fspadoni You marked the comments as resolved without doing any of the requested changes. Why?

fspadoni commented:
Sorry, I forgot to push the commit. Just did it
